homeassistant/components/config/config_entries.py

```diff
@@ -587,3 +639,61 @@ async def _async_matching_config_entries_json_fragments(
         )
         or (filter_is_not_helper and entry.domain not in integrations)
     ]
+
+
+@websocket_api.require_admin
+@websocket_api.websocket_command(
+    {
+        "type": "config_entries/subentries/list",
+        "entry_id": str,
+    }
+)
+@websocket_api.async_response
+async def config_subentry_list(
+    hass: HomeAssistant,
+    connection: websocket_api.ActiveConnection,
+    msg: dict[str, Any],
+) -> None:
+    """List subentries of a config entry."""
+    entry = get_entry(hass, connection, msg["entry_id"], msg["id"])
+    if entry is None:
+        return
+
+    result = {
+        "subentries": {
+            subentry_id: {"title": subentry.title}
+            for subentry_id, subentry in entry.subentries.items()
+        }
+    }
+    connection.send_result(msg["id"], result)
+
+
+@websocket_api.require_admin
+@websocket_api.websocket_command(
+    {
+        "type": "config_entries/subentries/delete",
+        "entry_id": str,
+        "subentry_id": str,
+    }
+)
+@websocket_api.async_response
+async def config_subentry_delete(
+    hass: HomeAssistant,
+    connection: websocket_api.ActiveConnection,
+    msg: dict[str, Any],
+) -> None:
+    """Delete a subentry of a config entry."""
+    entry = get_entry(hass, connection, msg["entry_id"], msg["id"])
+    if entry is None:
+        return
+
+    hass.config_entries.async_update_entry(
```

Comment:
I think that we need a special subentry removal function, so we can make sure that it's unloaded correctly and calls a subentry removal function if defined.

Reply:
I was also thinking about this.
However, this PR doesn't add a subentry setup function, so we can't really add a subentry removal function either?

I think we should leave it to the config entry to handle modified data, including added or removed subentries, via the config entry update listener for now.

Reply:
How will the config entry know what data got modified ? Now each integration needs to do a diff to see what was added/removed

Reply:
It's fine for this PR, but we should add this in a follow-up. Having lifecycle functions for subentries seems like it will be very useful for after the config entry has been setup.

Reply:
> How will the config entry know what data got modified ? Now each integration needs to do a diff to see what was added/removed

The common pattern is to simply make integrations reload themselves when settings are changed, I'd expect that would work well here too?

> Having lifecycle functions for subentries seems like it will be very useful for after the config entry has been setup.

An integration which could make use of an `async_setup_subentry` function should not care if that function is called right after or 1 week after the config entry was set up.

Hence, I think what would make most sense is to have `async_setup_subentry` + `async_unload_subentry`. The former would be called for all existing subentries after `async_setup_entry` and after a new subentry has been added. The latter would be called before `async_unload_entry` and after a subentry has been deleted.

Integrations who can't make use of this can just avoid implementing these additional entry points.